google/class-api-google-client.php

```diff
@@ -0,0 +1,264 @@
+<?php
+/**
+ * Class Yoast_Api_Google_Client
+ */
+class Yoast_Api_Google_Client extends Yoast_Google_Client {
+
+	/**
+	 * @var string
+	 */
+	protected $option_refresh_token;
+
+	/**
+	 * @var string
+	 */
+	protected $option_access_token;
+
+	/**
+	 * @var array
+	 */
+	protected $default_config = array();
+
+	/**
+	 * @var string
+	 */
+	protected $api_url;
+
+	/**
+	 * @var string
+	 */
+	protected $http_response_code;
+
+	/**
+	 * Initialize the config and refresh the token
+	 *
+	 * @param array $config
+	 */
+	public function __construct( $config ) {
+
+		parent::__construct();
+
+		// Initialize the config to set all properties properly.
+		$this->init_config( $config );
+
+		// Let's get an access token if we've got a refresh token.
+		$this->refresh_tokens();
+	}
+
+	/**
+	 * Authenticate the client. If $authorization_code is empty it will lead the user through the validation process of
+	 * Google. If set it will be get the access token for current session and save the refresh_token for future use
+	 *
+	 * @param mixed $authorization_code
+	 *
+	 * @return bool
+	 */
+	public function authenticate_client( $authorization_code = null ) {
+		static $has_retried;
+
+		// Authenticate client.
+		try {
+			$this->authenticate( $authorization_code );
+
+			// Get access response.
+			$response = $this->getAccessToken();
+
+			// Check if there is a response body.
+			if ( ! empty( $response ) ) {
+				$response = json_decode( $response );
+
+				if ( is_object( $response ) ) {
+					// Save the refresh token.
+					$this->save_refresh_token( $response->refresh_token );
+
+					return true;
+				}
+			}
+		} catch ( Yoast_Google_AuthException $exception ) {
+			// If there aren't any attempts before, try again and set attempts on true, to prevent further attempts.
+			if ( empty( $has_retried ) ) {
+				$has_retried = true;
+
+				return $this->authenticate_client( $authorization_code );
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Doing a request to the API
+	 *
+	 * @param string $target_request_url
+	 * @param string $request_method
+	 *
+	 * @return array
+	 */
+	public function do_request( $target_request_url, $request_method = 'GET' ) {
```

Comment:
please add a param `decode` (default false) so we can decode the response here instead of tons of other places where this is used.